core/src/main/scala/kafka/admin/ConfigCommand.scala

```diff
@@ -282,50 +282,69 @@ object ConfigCommand extends Config {
     val entityTypes = opts.options.valuesOf(opts.entityType).asScala
     if (entityTypes.size != 1)
       throw new IllegalArgumentException("Exactly one --entity-type must be specified with --bootstrap-server")
-    if (entityTypes.head != ConfigType.Broker)
-      throw new IllegalArgumentException(s"--zookeeper option must be specified for entity-type $entityTypes")
 
     try {
       if (opts.options.has(opts.alterOpt))
-        alterBrokerConfig(adminClient, opts, entityName)
+        alterBrokerConfig(adminClient, opts, entityTypes.head, entityName)
```

Comment:
Agreed. I need to manually play around with the command

Reply:
> opt.entityType lists the types of entities, since you added a new entity, you should add it there (so it will show up when we show the CLI help). I think this is necessary in few other places in ConfigCommandOptions class, like in add-config.

I don't think I understand - the only place I found that I can add it such that it comes out in the CLI help is the description of the --entity-type option:
```
val entityType = parser.accepts("entity-type", "Type of entity (topics/clients/users/brokers/broker-loggers)")
```
✅ added it there

Additionally, I played around with the command, here is some output:
```
./kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type broker-loggers
At least one of --entity-name or --entity-default must be specified with --bootstrap-server
```
```
./kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default
Error while executing config command with args '--bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default'
java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.InvalidRequestException: Broker id must not be empty
```
Hah - whoops. We don't have validation for describe with entity-default. Adding it ✅ 

--------------------------

Ugh, I get the following error locally:
```
[2019-06-24 10:50:23,401] ERROR [Admin Manager on Broker 0]: Error processing describe configs request for resource ConfigResource(type=BROKER_LOGGER, name='0') (kafka.server.AdminManager)
java.lang.NoClassDefFoundError: kafka/utils/Log4jController$
	at kafka.server.AdminManager.$anonfun$describeConfigs$1(AdminManager.scala:377)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:128)
	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at kafka.server.AdminManager.describeConfigs(AdminManager.scala:322)
	at kafka.server.KafkaApis.handleDescribeConfigsRequest(KafkaApis.scala:2298)
	at kafka.server.KafkaApis.handle(KafkaApis.scala:161)
	at kafka.server.KafkaRequestHandler.run(KafkaRequestHandler.scala:69)
	at java.base/java.lang.Thread.run(Thread.java:844)
```
Any advice on how to build the project such that this works? I had done `./gradlew clean` prior

Reply:
I didn't have any issue with the classes... I just build and it seemed to work. Weird! 

I did have other odd issues. I have one broker and `broker.id=0`

When I run:
`bin/kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default --entity-name 1` 
the command just hangs. Not sure if this is expected behavior but seems unfriendly?

When I run:
`bin/kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default --entity-name 0` I get all the dynamic configs for the broker, but not the loggers:
```
Configs for broker 0 are:
  org.apache.zookeeper.CreateMode=null sensitive=false synonyms={}
  kafka.coordinator.transaction.ProducerIdManager=null sensitive=false synonyms={}
  org.apache.kafka.common.utils.AppInfoParser=null sensitive=false synonyms={}
  kafka.utils.KafkaScheduler=null sensitive=false synonyms={}
  kafka.log.ProducerStateManager=null sensitive=false synonyms={}
...
```

Any idea?

Reply:
Regarding the CLI help. I think this is indeed the only place. I thought we may need something under "entity-name" and "entity-default", but looks like this isn't necessary.

The one thing I'm less sure of - how will users know that the right entity name for broker-logger is the broker-id (and not the name of a logger for instance)?

Reply:
**re: right entity name**
I thought about that too but I couldn't figure out whether adding broker id to the entity-name again would be better.
a) keep it as it is
```
parser.accepts("entity-name", "Name of entity (topic name/client id/user principal name/broker id)")
```
b) add broker id
```
parser.accepts("entity-name", "Name of entity (topic name/client id/user principal name/broker id/broker id)")
```


Here's another thing we can do - I'll add validation that the entity-name is an integer when a broker is given. 

**re: command hanging**
I got this behavior too when testing out a separate config - this is because you are giving it an invalid broker ID (1) and the admin client starts issuing requests. It eventually times out.

```
Error while executing config command with args '--bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default --entity-name 1'
java.util.concurrent.TimeoutException
	at org.apache.kafka.common.internals.KafkaFutureImpl$SingleWaiter.await(KafkaFutureImpl.java:108)
	at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:272)
	at kafka.admin.ConfigCommand$.brokerLoggerConfigs(ConfigCommand.scala:374)
	at kafka.admin.ConfigCommand$.describeBrokerConfig(ConfigCommand.scala:344)
	at kafka.admin.ConfigCommand$.processBrokerConfig(ConfigCommand.scala:288)
	at kafka.admin.ConfigCommand$.main(ConfigCommand.scala:82)
	at kafka.admin.ConfigCommand.main(ConfigCommand.scala)
```
Diving into it more there doesn't seem to be an easy way to fix this - when we can't find a node by its ID in our current metadata, we [request an update](https://github.com/apache/kafka/blob/5216839d8bcda5c00825a0f4be866a525d948da8/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L545) until the we either time out or the [provide method returns non-null](https://github.com/apache/kafka/blob/5216839d8bcda5c00825a0f4be866a525d948da8/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L897).
Testing this locally, I see my AdminClient creates 7.8 requests per second. This is quite the needless load on the cluster - it respects the retry.backoff.ms but maybe it is better to have an exponential backoff of sorts (like this [conference talk suggests](https://www.youtube.com/watch?v=NHNZab_AzZo&t=18m20s), for instance).
I opened https://issues.apache.org/jira/browse/KAFKA-8607 about it

**re: getting dynamic broker**
The output you gave is the broker loggers. I am getting the same output too. These are not the dynamic configs


Thanks for the thorough testing - I appreciate it. It's funny to see us hitting the same things  and I see we both agree CLI quality is very important